File: src-tauri/src/commands/system.rs
use arboard::Clipboard;
use image::{ImageBuffer, Rgba};
use std::time::{SystemTime, UNIX_EPOCH};

fn paths_to_strings(paths: Vec<std::path::PathBuf>) -> Vec<String> {
    paths
        .iter()
        .map(|path| path.to_string_lossy().to_string())
        .collect()
}

#[tauri::command]
pub fn open_file_dialog() -> Result<Vec<String>, String> {
    let files = rfd::FileDialog::new().pick_files().unwrap_or_default();
    Ok(paths_to_strings(files))
}

#[tauri::command]
pub fn open_folder_dialog() -> Result<Vec<String>, String> {
    let folder = rfd::FileDialog::new().pick_folder();
    Ok(folder
        .map(|path| vec![path.to_string_lossy().to_string()])
        .unwrap_or_default())
}

#[tauri::command]
pub fn tray_upload_select() -> Result<Vec<String>, String> {
    open_file_dialog()
}

#[tauri::command]
pub fn write_clipboard_text(text: String) -> Result<(), String> {
    let mut clipboard = Clipboard::new().map_err(|err| err.to_string())?;
    clipboard.set_text(text).map_err(|err| err.to_string())
}

#[tauri::command]
pub fn read_clipboard_image() -> Result<String, String> {
    let mut clipboard = Clipboard::new().map_err(|err| err.to_string())?;
    let image = clipboard
        .get_image()
        .map_err(|_| "clipboard image not found".to_string())?;

    let buffer: Vec<u8> = image.bytes.into_owned();
    let rgba =
        ImageBuffer::<Rgba<u8>, Vec<u8>>::from_raw(image.width as u32, image.height as u32, buffer)
            .ok_or_else(|| "failed to decode clipboard image".to_string())?;

    let ts = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map_err(|err| err.to_string())?
        .as_millis();
    let path = std::env::temp_dir().join(format!("cloud-pika-clipboard-{ts}.png"));

    rgba.save(&path).map_err(|err| err.to_string())?;
    Ok(path.to_string_lossy().to_string())
}

File: services/system.go
package services

import (
	"os"
	"time"

	"github.com/wailsapp/wails/v3/pkg/application"

	"github.com/goll/cloud-pika/internal/system"
)

// SystemService 暴露原生系统能力（对话框、剪贴板、剪贴板图片）。
type SystemService struct {
	app *application.App
}

// NewSystemService 基于 Wails 应用构建系统服务。
func NewSystemService(app *application.App) *SystemService {
	return &SystemService{app: app}
}

// OpenFileDialog 打开多选文件对话框，返回选中路径。
func (s *SystemService) OpenFileDialog() ([]string, error) {
	return s.app.Dialog.OpenFile().CanChooseFiles(true).PromptForMultipleSelection()
}

// OpenFolderDialog 打开文件夹选择对话框，返回选中目录（单元素切片）。
func (s *SystemService) OpenFolderDialog() ([]string, error) {
	path, err := s.app.Dialog.OpenFile().
		CanChooseDirectories(true).CanChooseFiles(false).PromptForSingleSelection()
	if err != nil || path == "" {
		return []string{}, err
	}
	return []string{path}, nil
}

// WriteClipboardText 写入文本到剪贴板。
func (s *SystemService) WriteClipboardText(text string) error {
	s.app.Clipboard.SetText(text)
	return nil
}

// ReadClipboardImage 读取剪贴板图片存临时 PNG，返回路径。
func (s *SystemService) ReadClipboardImage() (string, error) {
	return system.ReadClipboardImageToTemp(os.TempDir(), time.Now().UnixMilli())
}
